sim/core/spell_mod.go

```diff
@@ -0,0 +1,287 @@
+package core
+
+import (
+	"strconv"
+	"time"
+)
+
+/*
+SpellMod implementation.
+*/
+
+type SpellModConfig struct {
+	ClassMask  int64
+	Kind       SpellModType
+	School     SpellSchool
+	IntValue   int64
+	TimeValue  time.Duration
+	FloatValue float64
+}
+
+type SpellMod struct {
+	ClassMask      int64
+	Kind           SpellModType
+	School         SpellSchool
+	floatValue     float64
+	intValue       int64
+	Stacks         int32
+	timeValue      time.Duration
+	Apply          SpellModApply
+	Remove         SpellModRemove
+	IsActive       bool
+	AffectedSpells []*Spell
+}
+
+type SpellModApply func(mod *SpellMod, spell *Spell)
+type SpellModRemove func(mod *SpellMod, spell *Spell)
+type SpellModFunctions struct {
+	Apply  SpellModApply
+	Remove SpellModRemove
+}
+
+func buildMod(unit *Unit, config SpellModConfig) *SpellMod {
+	functions := spellModMap[config.Kind]
+	if functions == nil {
+		panic("SpellMod " + strconv.Itoa(int(config.Kind)) + " not implmented")
+	}
+
+	mod := &SpellMod{
+		ClassMask:  config.ClassMask,
+		Kind:       config.Kind,
+		School:     config.School,
+		floatValue: config.FloatValue,
+		intValue:   config.IntValue,
+		timeValue:  config.TimeValue,
+		Apply:      functions.Apply,
+		Remove:     functions.Remove,
+		IsActive:   false,
+		Stacks:     1,
+	}
+
+	for _, spell := range unit.Spellbook {
+		if shouldApply(spell, mod) {
+			mod.AffectedSpells = append(mod.AffectedSpells, spell)
+		}
+	}
+
+	unit.OnSpellRegistered(func(spell *Spell) {
+		if shouldApply(spell, mod) {
+			mod.AffectedSpells = append(mod.AffectedSpells, spell)
+
+			if mod.IsActive {
+				mod.Apply(mod, spell)
+			}
+		}
+	})
+
+	return mod
+}
+
+func (unit *Unit) AddStaticMod(config SpellModConfig) {
+	mod := buildMod(unit, config)
+	unit.ActiveSpellMods = append(unit.ActiveSpellMods, mod)
```

Comment:
where is this `unit.ActiveSpellMods` used? I dont see it anywhere so wondering why its needed. Also shouldnt the Activate/Deactivate call be responsible for adding the mod to that list anyway?

Reply:
It felt odd at first for the static's mod single point of reference to be through the callback handler of 'OnSpellRegistered'. Yet, you are right I never actually used it anywhere I believe. I'll clean that up